x/spec/types/spec.go

```diff
@@ -7,6 +7,97 @@ import (
 
 const minCU = 1
 
+type getterFunc func(index string) (Spec, bool)
+
+// ExpandSpec expands the "imports" field of the spec -if needed, recursively-
+// and adds to the current Spec those additional APIs from the imported Spec(s).
+// It takes as argument a getter function to fetch imported Spec(s) by name from
+// the keeper store, on demand. (Can't import the keeper module due to circular
+// dependencies)
+func (spec *Spec) ExpandSpec(getSpec getterFunc) (map[string]string, error) {
+	depends := map[string]bool{spec.Index: true}
+	details, err := spec.doExpandSpec(getSpec, depends, spec.Index)
+	return map[string]string{"spec": spec.Name, "imports": details}, err
+}
+
+// doExpandSpec performs the actual work and recusion for ExpandSpec above.
+func (spec *Spec) doExpandSpec(getSpec getterFunc, depends map[string]bool, details string) (string, error) {
+	if len(spec.Imports) == 0 {
+		return details, nil
+	}
+
+	var parents []Spec
+
+	// visual markers when import deepens
+	details += "->["
+
+	// recursion to get all parent specs
+	comma := ""
+	for _, index := range spec.Imports {
+		imported, found := getSpec(index)
+		// import of unknown Spec not allwoed
+		if !found {
+			details += fmt.Sprintf("%s%s(unknown)", comma, index)
+			return details, fmt.Errorf("imported spec unknown: %s", index)
+		}
+		// import of disabled Spec not allowed
+		if !imported.Enabled {
+			details += fmt.Sprintf("%s%s(disabled)", comma, index)
+			return details, fmt.Errorf("imported spec disabled: %s", index)
+		}
+
+		details += fmt.Sprintf("%s%s", comma, index)
+
+		// loop in the recursion not allowed
+		if _, found := depends[index]; found {
+			return details, fmt.Errorf("import loops not allowed for spec: %s", index)
+		}
+
+		depends[index] = true
+		details, err := imported.doExpandSpec(getSpec, depends, details)
+		if err != nil {
+			return details, err
+		}
+		delete(depends, index)
+
+		parents = append(parents, imported)
+		comma = ","
+	}
+
+	details += "]"
+
+	var mergedApis []ServiceApi
+	mergedApisMap := make(map[string]ServiceApi)
+
+	// collect all parents' Specs' APIs
+	for _, imported := range parents {
+		for _, api := range imported.Apis {
+			if api.Enabled {
+				// duplicate API(s) not allowed
```

Comment:
there's a case we might have conflict but it's overwritten anyway by the inheriting spec.Apis, there is no need to bail here, since those won't be used anyway

Reply:
Not sure I understand your comment - this is unrelated to overrides that the current Spec may (or may not) have.
The code detects cases where a Spec imports at once two or more Specs, which have APIs with the same name (but different details). The result of such import may be ambiguous (e.g. depending on the order), so we prohibit it.

Reply:
in the case of both functions importing the same API and then its ambiguous but discarded due to an override, then it didn't matter there was ambiguity and therefore this check can fail when in fact we could have continued

Reply:
Ok.